### DISCUSSION TITLE
Sending Teleport Logs to Datadog

### DISCUSSION THREAD

**Discussion by colinedwardwood** [gravitational/teleport#6615] [Show and tell]
## Install Agent
Install Agent - I used a single curl | bash provided by Datadog. I ran it on my Proxy/Auth node. This gave me metrics for system performance like CPU usage.

## Enable Log Collection
The Datadog Agent can directly grab the logs. Instructions for setup can be found [here](https://docs.datadoghq.com/agent/guide/agent-configuration-files/?tab=agentv6v7). Below are the steps I followed to enable Datadog to move the Teleport logs to the Datadog cloud.

### Update the Agent File
To enable log collection with an Agent running on your host, update the Agent’s main configuration file (datadog.yaml) with:

`logs_enabled: true`

The configuration file can be in different locations. The file location can be found [here](https://docs.datadoghq.com/agent/guide/agent-configuration-files/?tab=agentv6v7).

### Log Collection
Now Datadog will move logs, but which logs? We have to set it up to know about the Teleport logs. The general instructions are [here](https://docs.datadoghq.com/agent/logs/?tab=tailfiles).

Here are the steps I followed. In `/etc/datadog-agent/conf.d` I run:  
* `sudo mkdir teleport.d`
* `cd teleport.d`
* `sudo touch conf.yaml`
* `sudo vi conf.yaml`

Contents of `conf.yaml`.  

```
logs:
  - type: file
    path: "/var/lib/teleport/log/*/*.log"
    service: "Teleport"
    source: "Teleport"
```

**Readability of Log Files**  
By default, the Teleport directory `/var/lib/teleport` isn't readable and the Datadog-agent doesn't run as root. So we need to provide access.
`sudo chmod -R 755 /var/lib/teleport`
Datadog help on this issue is available [here](https://docs.datadoghq.com/logs/guide/log-collection-troubleshooting-guide/#permission-issues-tailing-log-files).

**Enable our Changes**  
Restart the datadog-agent with
`sudo systemctl restart datadog-agent`

**Confirm Logs Being Collected**
Check the Agent Status with `sudo datadog-agent status`.  
Look for Logs Agents Section which should show the Teleport directories.
e.g.  

```
==========
Logs Agent
==========

    Sending compressed logs in HTTPS to agent-http-intake.logs.datadoghq.com on port 443
    BytesSent: 0
    EncodedBytesSent: 28
    LogsProcessed: 0
    LogsSent: 0

  teleport
  --------
    - Type: file
      Path: /var/lib/teleport/log/*/*.log
      Status: OK
        31 files tailed out of 31 files matching
      Inputs: /var/lib/teleport/log/5e38ee26-52a6-454b-8914-cd836ebcd4aa/2021-04-20.00:00:00.log /var/lib/teleport/log/5e38ee26-52a6-454b-8914-cd836ebcd4aa/2021-03-22.00:00:00.log /var/lib/teleport/log/5e38ee26-52a6-454b-8914-cd836ebcd4aa/2021-04-22.00:00:00.log /var/lib/teleport/log/5e38ee26-52a6-454b-8914-cd836ebcd4aa/2021-03-31.00:00:00.log /var/lib/teleport/log/5e38ee26-52a6-454b-8914-cd836ebcd4aa/2021-03-26.00:00:00.log /var/lib/teleport/log/5e38ee26-52a6-454b-8914-cd836ebcd4aa/2021-03-23.00:00:00.log /var/lib/teleport/log/5e38ee26-52a6-454b-8914-cd836ebcd4aa/2021-03-16.00:00:00.log /var/lib/teleport/log/5e38ee26-52a6-454b-8914-cd836ebcd4aa/2021-03-09.00:00:00.log /var/lib/teleport/log/5e38ee26-52a6-454b-8914-cd836ebcd4aa/2021-04-12.00:00:00.log /var/lib/teleport/log/5e38ee26-52a6-454b-8914-cd836ebcd4aa/2021-04-07.00:00:00.log /var/lib/teleport/log/5e38ee26-52a6-454b-8914-cd836ebcd4aa/2021-04-06.00:00:00.log /var/lib/teleport/log/5e38ee26-52a6-454b-8914-cd836ebcd4aa/2021-04-09.00:00:00.log /var/lib/teleport/log/5e38ee26-52a6-454b-8914-cd836ebcd4aa/2021-03-04.00:00:00.log /var/lib/teleport/log/5e38ee26-52a6-454b-8914-cd836ebcd4aa/2021-04-26.00:00:00.log /var/lib/teleport/log/5e38ee26-52a6-454b-8914-cd836ebcd4aa/2021-04-23.00:00:00.log /var/lib/teleport/log/5e38ee26-52a6-454b-8914-cd836ebcd4aa/2021-04-16.00:00:00.log /var/lib/teleport/log/5e38ee26-52a6-454b-8914-cd836ebcd4aa/2021-03-24.00:00:00.log /var/lib/teleport/log/5e38ee26-52a6-454b-8914-cd836ebcd4aa/2021-03-08.00:00:00.log /var/lib/teleport/log/5e38ee26-52a6-454b-8914-cd836ebcd4aa/2021-03-05.00:00:00.log /var/lib/teleport/log/5e38ee26-52a6-454b-8914-cd836ebcd4aa/2021-04-19.00:00:00.log /var/lib/teleport/log/5e38ee26-52a6-454b-8914-cd836ebcd4aa/2021-04-10.00:00:00.log /var/lib/teleport/log/5e38ee26-52a6-454b-8914-cd836ebcd4aa/2021-03-30.00:00:00.log /var/lib/teleport/log/5e38ee26-52a6-454b-8914-cd836ebcd4aa/2021-04-14.00:00:00.log /var/lib/teleport/log/5e38ee26-52a6-454b-8914-cd836ebcd4aa/2021-04-08.00:00:00.log /var/lib/teleport/log/5e38ee26-52a6-454b-8914-cd836ebcd4aa/2021-03-25.00:00:00.log /var/lib/teleport/log/5e38ee26-52a6-454b-8914-cd836ebcd4aa/2021-04-01.00:00:00.log /var/lib/teleport/log/5e38ee26-52a6-454b-8914-cd836ebcd4aa/2021-03-11.00:00:00.log /var/lib/teleport/log/5e38ee26-52a6-454b-8914-cd836ebcd4aa/2021-04-27.00:00:00.log /var/lib/teleport/log/5e38ee26-52a6-454b-8914-cd836ebcd4aa/2021-04-15.00:00:00.log /var/lib/teleport/log/5e38ee26-52a6-454b-8914-cd836ebcd4aa/2021-04-13.00:00:00.log /var/lib/teleport/log/5e38ee26-52a6-454b-8914-cd836ebcd4aa/2021-04-21.00:00:00.log
      BytesRead: 0
      Average Latency (ms): 0
      24h Average Latency (ms): 0
```

**Comment by Saiteju1997** [comment]
Can we send teleport event handler logs to datadog????

**Reply by alexfornuto** [reply]
> Can we send teleport event handler logs to datadog????

Yes, by using the [Fluentd output plugin for Datadog](https://github.com/DataDog/fluent-plugin-datadog) - I'm working on a guide for this process as we speak...

**Reply by alexfornuto** [reply]
https://github.com/gravitational/teleport/pull/21950
